Core/include/Acts/EventData/VectorMultiTrajectory.hpp

```diff
@@ -0,0 +1,326 @@
+        return &m_index[istate].typeFlags;
+      default:
+        auto it = m_dynamic.find(key);
+        if (it == m_dynamic.end()) {
+          throw std::runtime_error("Unable to handle this component");
+        }
+        auto& col = it->second;
+        assert(col && "Dynamic column is null");
+        return col->get(istate);
+    }
+  }
+
+  template <typename T>
+  constexpr void addColumn_impl(const std::string& key) {
+    m_dynamic.insert({hashString(key), std::make_unique<DynamicColumn<T>>()});
+  }
+
+  constexpr bool hasColumn_impl(HashedString key) const {
+    using namespace Acts::HashedStringLiteral;
+    switch (key) {
+      case "predicted"_hash:
+      case "filtered"_hash:
+      case "smoothed"_hash:
+      case "calibrated"_hash:
+      case "jacobian"_hash:
+      case "projector"_hash:
+      case "previous"_hash:
+      case "uncalibrated"_hash:
+      case "calibratedSourceLink"_hash:
+      case "referenceSurface"_hash:
+      case "measdim"_hash:
+      case "chi2"_hash:
+      case "pathLength"_hash:
+      case "typeFlags"_hash:
+        return true;
+      default:
+        return m_dynamic.find(key) != m_dynamic.end();
+    }
+  }
+
+  // END INTERFACE
+
+  struct DynamicColumnBase {
+    virtual ~DynamicColumnBase() = 0;
+
+    virtual std::any get(size_t i) = 0;
+    virtual std::any get(size_t i) const = 0;
+
+    virtual void add() = 0;
+    virtual void clear() = 0;
+
+    virtual std::unique_ptr<DynamicColumnBase> clone() const = 0;
+  };
+
+  template <typename T>
+  struct DynamicColumn : public VectorMultiTrajectory::DynamicColumnBase {
+    ~DynamicColumn() override = default;
+
+    std::any get(size_t i) override {
+      assert(i < m_vector.size() && "DynamicColumn out of bounds");
+      return &m_vector[i];
+    }
+
+    std::any get(size_t i) const override {
+      assert(i < m_vector.size() && "DynamicColumn out of bounds");
+      return &m_vector[i];
+    }
+
+    void add() override { m_vector.emplace_back(); }
+    void clear() override { m_vector.clear(); }
+
+    std::unique_ptr<DynamicColumnBase> clone() const override {
+      return std::make_unique<DynamicColumn<T>>(*this);
+    }
+
+    std::vector<T> m_vector;
+  };
+
+  /// index to map track states to the corresponding
+  std::vector<IndexData> m_index;
+  std::vector<IndexType> m_previous;
+  std::vector<typename detail_lt::Types<eBoundSize>::Coefficients> m_params;
+  std::vector<typename detail_lt::Types<eBoundSize>::Covariance> m_cov;
+  std::vector<typename detail_lt::Types<MeasurementSizeMax>::Coefficients>
+      m_meas;
+  std::vector<typename detail_lt::Types<MeasurementSizeMax>::Covariance>
+      m_measCov;
+  std::vector<typename detail_lt::Types<eBoundSize>::Covariance> m_jac;
+  std::vector<const SourceLink*> m_sourceLinks;
+  std::vector<ProjectorBitset> m_projectors;
+
+  // owning vector of shared pointers to surfaces
+  // @TODO: This might be problematic when appending a large number of surfaces
```

Comment:
todo?

Reply:
This is more of a consideration i guess. Removed the TODO label from the comment.